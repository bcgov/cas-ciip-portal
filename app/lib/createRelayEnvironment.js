import {Environment, Network, RecordSource, Store} from 'relay-runtime';
import fetch from 'isomorphic-unfetch';

let relayEnvironment = null;

// Define a function that fetches the results of an operation (query/mutation/etc)
// and returns its results as a Promise:
function fetchQuery(operation, variables, cacheConfig, uploadables) {
  // Point client to same host as the graphql server
  const relayServer =
    typeof window !== 'undefined' ? '' : process.env.RELAY_SERVER;
  // Return fetch(`${relayServer}/graphql`, {
  return fetch('http://localhost:3004/graphql', {
    // Todo: import from constants
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }, // Todo: Add authentication and other headers here
    body: JSON.stringify({
      query: operation.text, // GraphQL text from input
      variables
    })
  }).then(response => {
    // Console.log('CreateRelayEnvironment.js > fetchQuery() x', response.json());
    return response.json();
  });
}

export default function initEnvironment({records = {}} = {}) {
  // Create a network layer from the fetch function
  const network = Network.create(fetchQuery);
  const store = new Store(new RecordSource(records));

  // Make sure to create a new Relay environment for every server-side request so that data
  // isn't shared between connections (which would be bad)
  if (typeof window === 'undefined') {
    return new Environment({
      network,
      store
    });
  }

  // Reuse Relay environment on client-side
  if (!relayEnvironment) {
    relayEnvironment = new Environment({
      network,
      store
    });
  }

  return relayEnvironment;
}
