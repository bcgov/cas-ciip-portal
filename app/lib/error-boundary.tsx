import React, {Component} from 'react';
import {DefaultLayoutComponent} from 'layouts/default-layout';
import {Alert} from 'react-bootstrap';
import getConfig from 'next/config';

class ErrorBoundary extends Component {
  state = {hasError: false, error: null, errorInfo: null};

  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI.
    console.error(error);
    return {hasError: true};
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    // logErrorToMyService(error, errorInfo);
    this.setState({error, errorInfo});
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      const feedbackUrl = getConfig()?.publicRuntimeConfig.FEEDBACK_SITE_URL;
      // You can render any custom fallback UI
      return (
        <DefaultLayoutComponent
          needsSession={false}
          needsUser={false}
          session={null}
        >
          <Alert variant="danger">
            <Alert.Heading>An unexpected error has occured</Alert.Heading>
            <p>
              Please consider reporting this error on our{' '}
              <Alert.Link href={feedbackUrl}>feedback site</Alert.Link>, by
              either creating a new post or commenting on an existing post if
              this error was already reported. Copying the error details below
              when submitting a bug report will help us understand what
              happened.
            </p>
          </Alert>

          <h3>Error details</h3>
          <pre>
            <code>
              {this.state.error?.toString()}
              {process.env.NODE_ENV !== 'production' &&
                this.state.errorInfo?.componentStack}
            </code>
          </pre>
        </DefaultLayoutComponent>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
