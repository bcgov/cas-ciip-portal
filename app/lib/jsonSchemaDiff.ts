import diff from "deep-diff";

const addObjectToDiffMap = (path, lhs, rhs, idDiffMap) => {
  if (typeof lhs !== "object" && typeof rhs !== "object") {
    idDiffMap[path] = { lhs: lhs ?? null, rhs: rhs ?? null };
    return;
  }
  if (typeof lhs === "object" || typeof rhs === "object") {
    const keys = new Set<string>();

    if (lhs) Object.keys(lhs).forEach((k) => keys.add(k));
    if (rhs) Object.keys(rhs).forEach((k) => keys.add(k));

    keys.forEach((k) =>
      addObjectToDiffMap(`${path}_${k}`, lhs?.[k], rhs?.[k], idDiffMap)
    );
  }
};

/**
 *
 * @param formData the form data to display when not diffing
 * @param formIdPrefix
 * @param lhs
 * @param rhs
 * @returns An object with the formData to display, and the idDiffMap, containing the field-level diffs
 */
const useJsonSchemaDiff = (
  formData: any,
  formIdPrefix: string,
  lhs?: any,
  rhs?: any
) => {
  const idDiffMap: Record<string, { lhs: any; rhs: any }> = {};
  const differences = diff(lhs, rhs);
  let newFormData = Array.isArray(rhs) ? [...rhs] : { ...rhs };

  if (differences) {
    // Populate the diffPathArray and diffArray
    differences.forEach((difference) => {
      if (difference.kind === "A") {
        // Array differences
        const arrayElementPath = [
          formIdPrefix,
          ...(difference.path ?? []),
          difference.index,
        ].join("_");
        if (difference.item.kind === "N") {
          // Added an element to the array
          Object.keys(difference.item.rhs).forEach((key) => {
            idDiffMap[`${arrayElementPath}_${key}`] = {
              lhs: null,
              rhs: difference.item.rhs[key],
            };
          });
        } else if (difference.item.kind === "D") {
          // Deleted an element from the array
          Object.keys(difference.item.lhs).forEach((key) => {
            idDiffMap[`${arrayElementPath}_${key}`] = {
              lhs: difference.item.lhs[key],
              rhs: null,
            };
          });
          if (Array.isArray(newFormData) && !difference.path) {
            newFormData = [
              ...newFormData.slice(0, difference.index),
              difference.item.lhs,
              ...newFormData.slice(difference.index),
            ];
          }
        }
      } else if (difference.path) {
        const { lhs, rhs } = difference;
        addObjectToDiffMap(
          [formIdPrefix, ...difference.path].join("_"),
          lhs,
          rhs,
          idDiffMap
        );
      }
    });
  }

  return { formData: newFormData, idDiffMap };
};

export default useJsonSchemaDiff;
