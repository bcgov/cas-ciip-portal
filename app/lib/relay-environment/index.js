import 'isomorphic-fetch'

export const { initEnvironment, createEnvironment } = (typeof window ===
'undefined'
  ? require('./server')
  : require('./client')
).default
