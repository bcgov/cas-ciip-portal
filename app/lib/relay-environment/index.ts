/**
 * We need to polyfill `fetch()` for react-relay-network-modern
 * in both the server and client environments.
 * @see https://github.com/relay-tools/react-relay-network-modern
 * @see https://fetch.spec.whatwg.org/
 **/
import "isomorphic-fetch";
// ^this has side-effects since it's a global polyfill!
// TODO: does this conflict with Next.js use of `unfetch` internally?

export const { initEnvironment, createEnvironment } = (typeof window ===
"undefined"
  ? require("./server")
  : require("./client")
).default;
