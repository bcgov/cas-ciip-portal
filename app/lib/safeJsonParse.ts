// Function takes a string as a parameter & returns a parsed JSON object if valid JSON, {} if not valid.
// Prevents app crash on invalid querystring.
export default function safeJsonParse(toParse: string) {
  try {
    return JSON.parse(toParse);
  } catch (e) {
    return {};
  }
}
