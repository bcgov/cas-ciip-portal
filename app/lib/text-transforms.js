function capitalize(input) {
  if (typeof input !== "string" && !Number.isFinite(input)) return "";
  const str = `${input}`;
  return `${str.slice(0, 1).toUpperCase()}${str.slice(1).toLowerCase()}`;
}

function getUserFriendlyStatusLabel(status) {
  const s = status === "REQUESTED_CHANGES" ? "CHANGES REQUESTED" : status;
  return `${capitalize(s).replace(/_/g, " ")}`;
}

export { getUserFriendlyStatusLabel, capitalize };
