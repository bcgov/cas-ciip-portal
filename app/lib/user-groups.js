const GROUP_META = require('../data/groups.json');

const getAllGroups = () => {
  return Object.keys(GROUP_META);
};

const compactGroups = (groups) => {
  const allDefinedGroups = getAllGroups();

  return groups.filter((group) => allDefinedGroups.includes(group));
};

const getPriorityGroupData = (validGroups) => {
  // Find the highest priority group
  let priorityGroup = {
    name: 'Guest',
    ...GROUP_META.Guest
  };

  for (let x = 0, len = validGroups.length; x < len; x++) {
    const name = validGroups[x];
    const curr = GROUP_META[name];

    if (curr.priority < priorityGroup.priority) {
      priorityGroup = {...curr, name};
    }
  }

  return priorityGroup;
};

const getPriorityGroup = (groupNames) => {
  const validGroups = compactGroups(groupNames);
  const priorityGroupData = getPriorityGroupData(validGroups);

  return priorityGroupData.name;
};

const getUserGroupLandingRoute = (groupNames) => {
  const validGroups = compactGroups(groupNames);
  const priorityGroupData = getPriorityGroupData(validGroups);

  return priorityGroupData.path;
};

module.exports = {
  getUserGroupLandingRoute,
  getPriorityGroup,
  compactGroups,
  getAllGroups
};
