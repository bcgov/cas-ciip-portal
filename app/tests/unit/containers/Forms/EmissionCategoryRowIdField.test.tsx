import React from 'react';
import {shallow} from 'enzyme';
import {
  EmissionCategoryRowIdFieldComponent,
  Props
} from 'containers/Forms/EmissionCategoryRowIdField';
import {getDefaultRegistry} from 'react-jsonschema-form/lib/utils';

describe('The EmissionCategoryRowIdField component', () => {
  it('injects the category ids and names in the json schema', () => {
    const idSchema: any = {$id: 'emissionCategoryRowId'}; // TODO: revise type after react-jsonschema-form v2 is used
    const props: Props = {
      name: 'emissionCategoryRowId',
      schema: {
        type: 'number',
        title: 'emission category'
      },
      uiSchema: {},
      idSchema,
      formData: 1,
      errorSchema: null,
      onChange: jest.fn,
      onBlur: jest.fn,
      registry: getDefaultRegistry(),
      formContext: null,
      autofocus: false,
      disabled: false,
      readonly: false,
      required: true,
      query: {
        ' $refType': 'EmissionCategoryRowIdField_query',
        allEmissionCategories: {
          edges: [
            {
              node: {
                rowId: 1,
                displayName: 'cat A'
              }
            },
            {
              node: {
                rowId: 2,
                displayName: 'cat B'
              }
            }
          ]
        }
      }
    };
    const renderedField = shallow(
      <EmissionCategoryRowIdFieldComponent {...props} />
    );
    expect(renderedField).toMatchSnapshot();
    const fieldSchemaProp = renderedField.prop('schema');
    expect(fieldSchemaProp.enum).toHaveLength(2);
    expect(fieldSchemaProp.enumNames).toHaveLength(2);
  });
});
