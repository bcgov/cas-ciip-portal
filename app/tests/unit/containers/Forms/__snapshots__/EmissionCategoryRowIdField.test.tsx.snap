// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The EmissionCategoryRowIdField component injects the category ids and names in the json schema 1`] = `
<StringField
  autofocus={false}
  disabled={false}
  errorSchema={null}
  formContext={null}
  formData={1}
  idSchema={
    Object {
      "$id": "emissionCategoryRowId",
    }
  }
  name="emissionCategoryRowId"
  onBlur={[Function]}
  onChange={[Function]}
  readonly={false}
  registry={
    Object {
      "definitions": Object {},
      "fields": Object {
        "AnyOfField": [Function],
        "ArrayField": [Function],
        "BooleanField": [Function],
        "DescriptionField": [Function],
        "NullField": [Function],
        "NumberField": [Function],
        "ObjectField": [Function],
        "OneOfField": [Function],
        "SchemaField": [Function],
        "StringField": [Function],
        "TitleField": [Function],
        "UnsupportedField": [Function],
      },
      "formContext": Object {},
      "widgets": Object {
        "AltDateTimeWidget": [Function],
        "AltDateWidget": [Function],
        "BaseInput": [Function],
        "CheckboxWidget": [Function],
        "CheckboxesWidget": [Function],
        "ColorWidget": [Function],
        "DateTimeWidget": [Function],
        "DateWidget": [Function],
        "EmailWidget": [Function],
        "FileWidget": [Function],
        "HiddenWidget": [Function],
        "PasswordWidget": [Function],
        "RadioWidget": [Function],
        "RangeWidget": [Function],
        "SelectWidget": [Function],
        "TextWidget": [Function],
        "TextareaWidget": [Function],
        "URLWidget": [Function],
        "UpDownWidget": [Function],
      },
    }
  }
  required={true}
  schema={
    Object {
      "enum": Array [
        1,
        2,
      ],
      "enumNames": Array [
        "cat A",
        "cat B",
      ],
      "title": "emission category",
      "type": "number",
    }
  }
  uiSchema={Object {}}
/>
`;
